File: flask_app/app/models/product.py
from app.models.base import BaseModel, db

class Product(BaseModel):
    __tablename__ = 'product'

    name = db.Column(db.String(100), unique=True, nullable=False)
    description = db.Column(db.Text, nullable=False)
    category_id  = db.Column(db.String(36), nullable=False)
    stock_quantity = db.Column(db.Integer, nullable=False, default=0)
    discount_id  = db.Column(db.String(36), nullable=True)
    price = db.Column(db.Float, nullable=False, default=0)
    is_membership_exclusive = db.Column(db.Boolean, default=False)
    min_tier_required = db.Column(db.Integer, nullable=False, default=1)
    
    @property
    def is_available(self):
        return self.stock_quantity > 0
    
    def __repr__(self):
        return f"<Product {self.name}>"

File: java_membership_api/src/main/java/com/example/membership_api/model/Product.java
package com.example.membership_api.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Table;
import jakarta.persistence.Transient;
import lombok.Data;
import lombok.EqualsAndHashCode;

@Data
@Entity
@Table(name = "product")
@EqualsAndHashCode(callSuper = true)
public class Product extends BaseModel{

    @Column(unique = true, nullable = false, length = 100)
    private String name;

    @Column(nullable = false)
    private String description;

    @Column(name = "category_id", nullable = false)
    private String categoryId;

    @Column(name = "stock_quantity", nullable = false)
    private Integer stockQuantity = 0;

    @Column(name = "discount_id")
    private String discountId;

    @Column(nullable = false)
    private Double price = 0.0;

    @Column(name = "is_membership_exclusive", nullable = false)
    private Boolean isMembershipExclusive = false;

    @Column(name = "min_tier_required", nullable = false)
    private Integer minTierRequired = 1;

    @Transient
    public boolean isAvailable() {
        return stockQuantity != null && stockQuantity > 0;
    }
}
